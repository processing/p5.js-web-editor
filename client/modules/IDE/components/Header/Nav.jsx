import React, { useContext } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { sortBy } from 'lodash';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import MediaQuery from 'react-responsive';
import NavDropdownMenu from '../../../../components/Nav/NavDropdownMenu';
import NavMenuItem from '../../../../components/Nav/NavMenuItem';
import { availableLanguages, languageKeyToLabel } from '../../../../i18n';
import getConfig from '../../../../utils/getConfig';
import { showToast } from '../../actions/toast';
import { setLanguage } from '../../actions/preferences';
import NavBar from '../../../../components/Nav/NavBar';
import CaretLeftIcon from '../../../../images/left-arrow.svg';
import LogoIcon from '../../../../images/p5js-logo-small.svg';
import { selectRootFile } from '../../selectors/files';
import { selectSketchPath } from '../../selectors/project';
import { metaKey, metaKeyName } from '../../../../utils/metaKey';
import { useSketchActions } from '../../hooks';
import { getAuthenticated, getIsUserOwner } from '../../selectors/users';
import { cloneProject } from '../../actions/project';
import {
  newFile,
  newFolder,
  showKeyboardShortcutModal,
  startSketch,
  stopSketch
} from '../../actions/ide';
import { logoutUser } from '../../../User/actions';
import { CmControllerContext } from '../../pages/IDEView';
import MobileNav from './MobileNav';

const Nav = ({ layout }) => (
  <MediaQuery minWidth={770}>
    {(matches) =>
      matches ? (
        <NavBar>
          <LeftLayout layout={layout} />
          <UserMenu />
        </NavBar>
      ) : (
        <MobileNav />
      )
    }
  </MediaQuery>
);

Nav.propTypes = {
  layout: PropTypes.oneOf(['dashboard', 'project'])
};

Nav.defaultProps = {
  layout: 'project'
};

const LeftLayout = (props) => {
  switch (props.layout) {
    case 'dashboard':
      return <DashboardMenu />;
    case 'project':
    default:
      return <ProjectMenu />;
  }
};

LeftLayout.propTypes = {
  layout: PropTypes.oneOf(['dashboard', 'project'])
};

LeftLayout.defaultProps = {
  layout: 'project'
};

const UserMenu = () => {
  const isLoginEnabled = getConfig('LOGIN_ENABLED');
  const isAuthenticated = useSelector(getAuthenticated);

  if (isLoginEnabled && isAuthenticated) {
    return <AuthenticatedUserMenu />;
  } else if (isLoginEnabled && !isAuthenticated) {
    return <UnauthenticatedUserMenu />;
  }

  return null;
};

const DashboardMenu = () => {
  const { t } = useTranslation();
  const editorLink = useSelector(selectSketchPath);
  return (
    <ul className="nav__items-left">
      <li className="nav__item-logo">
        <LogoIcon
          role="img"
          aria-label={t('Common.p5logoARIA')}
          focusable="false"
          className="svg__logo"
        />
      </li>
      <li className="nav__item nav__item--no-icon">
        <Link to={editorLink} className="nav__back-link">
          <CaretLeftIcon
            className="nav__back-icon"
            focusable="false"
            aria-hidden="true"
          />
          <span className="nav__item-header">{t('Nav.BackEditor')}</span>
        </Link>
      </li>
    </ul>
  );
};

const ProjectMenu = () => {
  const isUserOwner = useSelector(getIsUserOwner);
  const project = useSelector((state) => state.project);
  const user = useSelector((state) => state.user);

  const isUnsaved = !project?.id;

  const rootFile = useSelector(selectRootFile);

  const cmRef = useContext(CmControllerContext);

  const dispatch = useDispatch();

  const { t } = useTranslation();
  const {
    newSketch,
    saveSketch,
    downloadSketch,
    shareSketch
  } = useSketchActions();

  const replaceCommand =
    metaKey === 'Ctrl' ? `${metaKeyName}+H` : `${metaKeyName}+‚å•+F`;

  return (
    <ul className="nav__items-left">
      <li className="nav__item-logo">
        <LogoIcon
          role="img"
          aria-label={t('Common.p5logoARIA')}
          focusable="false"
          className="svg__logo"
        />
      </li>
      <NavDropdownMenu id="file" title={t('Nav.File.Title')}>
        <NavMenuItem onClick={newSketch}>{t('Nav.File.New')}</NavMenuItem>
        <NavMenuItem
          hideIf={
            !getConfig('LOGIN_ENABLED') || (project?.owner && !isUserOwner)
          }
          onClick={() => saveSketch(cmRef.current)}
        >
          {t('Common.Save')}
          <span className="nav__keyboard-shortcut">{metaKeyName}+S</span>
        </NavMenuItem>
        <NavMenuItem
          hideIf={isUnsaved || !user.authenticated}
          onClick={() => dispatch(cloneProject())}
        >
          {t('Nav.File.Duplicate')}
        </NavMenuItem>
        <NavMenuItem hideIf={isUnsaved} onClick={shareSketch}>
          {t('Nav.File.Share')}
        </NavMenuItem>
        <NavMenuItem hideIf={isUnsaved} onClick={downloadSketch}>
          {t('Nav.File.Download')}
        </NavMenuItem>
        <NavMenuItem
          hideIf={!user.authenticated}
          href={`/${user.username}/sketches`}
        >
          {t('Nav.File.Open')}
        </NavMenuItem>
        <NavMenuItem
          hideIf={
            !getConfig('UI_COLLECTIONS_ENABLED') ||
            !user.authenticated ||
            isUnsaved
          }
          href={`/${user.username}/sketches/${project?.id}/add-to-collection`}
        >
          {t('Nav.File.AddToCollection')}
        </NavMenuItem>
        <NavMenuItem
          hideIf={!getConfig('EXAMPLES_ENABLED')}
          href="/p5/sketches"
        >
          {t('Nav.File.Examples')}
        </NavMenuItem>
      </NavDropdownMenu>
      <NavDropdownMenu id="edit" title={t('Nav.Edit.Title')}>
        <NavMenuItem onClick={cmRef.current?.tidyCode}>
          {t('Nav.Edit.TidyCode')}
          <span className="nav__keyboard-shortcut">
            {metaKeyName}+{'\u21E7'}+F
          </span>
        </NavMenuItem>
        <NavMenuItem onClick={cmRef.current?.showFind}>
          {t('Nav.Edit.Find')}
          <span className="nav__keyboard-shortcut">{metaKeyName}+F</span>
        </NavMenuItem>
        <NavMenuItem onClick={cmRef.current?.showReplace}>
          {t('Nav.Edit.Replace')}
          <span className="nav__keyboard-shortcut">{replaceCommand}</span>
        </NavMenuItem>
      </NavDropdownMenu>
      <NavDropdownMenu id="sketch" title={t('Nav.Sketch.Title')}>
        <NavMenuItem onClick={() => dispatch(newFile(rootFile.id))}>
          {t('Nav.Sketch.AddFile')}
        </NavMenuItem>
        <NavMenuItem onClick={() => dispatch(newFolder(rootFile.id))}>
          {t('Nav.Sketch.AddFolder')}
        </NavMenuItem>
        <NavMenuItem onClick={() => dispatch(startSketch())}>
          {t('Nav.Sketch.Run')}
          <span className="nav__keyboard-shortcut">{metaKeyName}+Enter</span>
        </NavMenuItem>
        <NavMenuItem onClick={() => dispatch(stopSketch())}>
          {t('Nav.Sketch.Stop')}
          <span className="nav__keyboard-shortcut">
            {'\u21E7'}+{metaKeyName}+Enter
          </span>
        </NavMenuItem>
      </NavDropdownMenu>
      <NavDropdownMenu id="help" title={t('Nav.Help.Title')}>
        <NavMenuItem onClick={() => dispatch(showKeyboardShortcutModal())}>
          {t('Nav.Help.KeyboardShortcuts')}
        </NavMenuItem>
        <NavMenuItem href="https://p5js.org/reference/">
          {t('Nav.Help.Reference')}
        </NavMenuItem>
        <NavMenuItem href="/about">{t('Nav.Help.About')}</NavMenuItem>
      </NavDropdownMenu>
    </ul>
  );
};

const LanguageMenu = () => {
  const language = useSelector((state) => state.preferences.language);
  const dispatch = useDispatch();

  function handleLangSelection(event) {
    dispatch(setLanguage(event.target.value));
    dispatch(showToast('Toast.LangChange'));
  }

  return (
    <NavDropdownMenu id="lang" title={languageKeyToLabel(language)}>
      {sortBy(availableLanguages).map((key) => (
        // eslint-disable-next-line react/jsx-no-bind
        <NavMenuItem key={key} value={key} onClick={handleLangSelection}>
          {languageKeyToLabel(key)}
        </NavMenuItem>
      ))}
    </NavDropdownMenu>
  );
};

const UnauthenticatedUserMenu = () => {
  const { t } = useTranslation();
  return (
    <ul className="nav__items-right" title="user-menu">
      {getConfig('TRANSLATIONS_ENABLED') && <LanguageMenu />}
      <li className="nav__item">
        <Link to="/login" className="nav__auth-button">
          <span className="nav__item-header" title="Login">
            {t('Nav.Login')}
          </span>
        </Link>
      </li>
      <span className="nav__item-or">{t('Nav.LoginOr')}</span>
      <li className="nav__item">
        <Link to="/signup" className="nav__auth-button">
          <span className="nav__item-header" title="SignUp">
            {t('Nav.SignUp')}
          </span>
        </Link>
      </li>
    </ul>
  );
};

const AuthenticatedUserMenu = () => {
  const username = useSelector((state) => state.user.username);

  const { t } = useTranslation();
  const dispatch = useDispatch();

  return (
    <ul className="nav__items-right" title="user-menu">
      {getConfig('TRANSLATIONS_ENABLED') && <LanguageMenu />}
      <NavDropdownMenu
        id="account"
        title={
          <span>
            {t('Nav.Auth.Hello')}, {username}!
          </span>
        }
      >
        <NavMenuItem href={`/${username}/sketches`}>
          {t('Nav.Auth.MySketches')}
        </NavMenuItem>
        <NavMenuItem
          href={`/${username}/collections`}
          hideIf={!getConfig('UI_COLLECTIONS_ENABLED')}
        >
          {t('Nav.Auth.MyCollections')}
        </NavMenuItem>
        <NavMenuItem href={`/${username}/assets`}>
          {t('Nav.Auth.MyAssets')}
        </NavMenuItem>
        <NavMenuItem href="/account">{t('Preferences.Settings')}</NavMenuItem>
        <NavMenuItem onClick={() => dispatch(logoutUser())}>
          {t('Nav.Auth.LogOut')}
        </NavMenuItem>
      </NavDropdownMenu>
    </ul>
  );
};

export default Nav;
