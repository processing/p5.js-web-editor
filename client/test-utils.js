/**
 * This file re-exports @testing-library but ensures that
 * any calls to render have translations and theme available.
 *
 * This means tested components will be able to call
 * `t()` and have the translations of the default
 * language also components will be able to call
 * `prop()` and have the theming of the default theme.
 *
 * For i18n see: https://react.i18next.com/misc/testing#testing-without-stubbing
 */

// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
import React from 'react';
import PropTypes from 'prop-types';
import { createStore } from 'redux';
import { Provider } from 'react-redux';

import { I18nextProvider } from 'react-i18next';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';

import i18n from './i18n-test';
import rootReducer from './reducers';
import ThemeProvider from './modules/App/components/ThemeProvider';
import theme, { Theme } from './theme';

// re-export everything
// eslint-disable-next-line import/no-extraneous-dependencies
export * from '@testing-library/react';

const Providers = ({ children }) => (
  // eslint-disable-next-line react/jsx-filename-extension
  <StyledThemeProvider theme={{ ...theme[Theme.light] }}>
    <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
  </StyledThemeProvider>
);

Providers.propTypes = {
  children: PropTypes.element.isRequired
};

function reduxRender(
  ui,
  {
    initialState,
    store = createStore(rootReducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <I18nextProvider i18n={i18n}>
        <Provider store={store}>
          <ThemeProvider>{children}</ThemeProvider>
        </Provider>
      </I18nextProvider>
    );
  }

  Wrapper.propTypes = {
    children: PropTypes.element.isRequired
  };

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

const customRender = (ui, options) =>
  render(ui, { wrapper: Providers, ...options });

// override render method
export { customRender as render, reduxRender };
