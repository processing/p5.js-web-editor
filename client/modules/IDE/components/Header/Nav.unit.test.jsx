import React from 'react';
import { reduxRender } from '../../../../test-utils';

import Nav from './Nav';

jest.mock('../../../../utils/generateRandomName');

describe('Nav', () => {
  it('renders editor version for desktop', () => {
    const { asFragment } = reduxRender(<Nav />, { mobile: false });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders editor version for mobile', () => {
    const { asFragment } = reduxRender(<Nav />, { mobile: true });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders dashboard version for desktop', () => {
    const { asFragment } = reduxRender(<Nav layout="dashboard" />, {
      mobile: false
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders dashboard version for mobile', () => {
    const { asFragment } = reduxRender(<Nav layout="dashboard" />, {
      mobile: true
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
