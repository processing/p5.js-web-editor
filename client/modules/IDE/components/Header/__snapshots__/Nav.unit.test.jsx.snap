// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav renders dashboard version for desktop 1`] = `
<DocumentFragment>
  <header>
    <nav
      class="nav"
    >
      <ul
        class="nav__items-left"
      >
        <li
          class="nav__item-logo"
        >
          <test-file-stub
            aria-label="p5.js Logo"
            classname="svg__logo"
            focusable="false"
            role="img"
          />
        </li>
        <li
          class="nav__item nav__item--no-icon"
        >
          <a
            class="nav__back-link"
            href="/"
          >
            <test-file-stub
              aria-hidden="true"
              classname="nav__back-icon"
              focusable="false"
            />
            <span
              class="nav__item-header"
            >
              Back to Editor
            </span>
          </a>
        </li>
      </ul>
    </nav>
  </header>
</DocumentFragment>
`;

exports[`Nav renders dashboard version for mobile 1`] = `
<DocumentFragment>
  .c4.c4.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #333;
  cursor: pointer;
  border: none;
  line-height: 1;
}

.c4.c4.c4 svg * {
  fill: #333;
}

.c4.c4.c4:disabled {
  cursor: not-allowed;
}

.c4.c4.c4 > * + * {
  margin-left: 0.6666666666666666rem;
}

.c5 {
  width: 4rem;
}

.c5 > svg {
  width: 100%;
  height: 100%;
}

.c0 {
  background: #F0F0F0;
  color: #333;
  padding: 0.6666666666666666rem 0;
  gap: 1.1666666666666667rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.8333333333333334rem;
  box-shadow: #00000030 0px 2px 8px 0px;
  z-index: 10;
}

.c1 {
  width: 2.3333333333333335rem;
  aspect-ratio: 1;
  margin-left: 1.1666666666666667rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 > svg {
  width: 100%;
  height: 100%;
}

.c1 > svg > path {
  fill: #ed225d;
  stroke: #ed225d;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.16666666666666666rem;
}

.c2 * {
  padding: 0;
  margin: 0;
}

.c2 > h5 {
  font-size: 1.0833333333333333rem;
  font-weight: normal;
}

.c3 {
  margin-left: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 ul.opened {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  opacity: 1;
  max-width: 90vw;
  max-height: 90vh;
  overflow-y: auto;
}

.c3 > div {
  position: relative;
}

.c3 > div ul {
  foreground: #000;
  background: #F0F0F0;
  border: #A6A6A6;
  position: absolute;
  overflow: hidden;
  z-index: 10;
  right: 10px;
  -webkit-transform: translateX(0.5rem);
  -ms-transform: translateX(0.5rem);
  transform: translateX(0.5rem);
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  opacity: 0;
  -webkit-transform-origin: top;
  -ms-transform-origin: top;
  transform-origin: top;
  -webkit-transition: opacity 100ms,-webkit-transform 100ms 200ms;
  -webkit-transition: opacity 100ms,transform 100ms 200ms;
  transition: opacity 100ms,transform 100ms 200ms;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.16666666666666666rem;
  box-shadow: rgba(60,64,67,0.3) 0px 1px 2px 0px, rgba(60,64,67,0.15) 0px 1px 3px 1px;
  min-width: 10rem;
  border-radius: 0.16666666666666666rem;
}

.c3 > div ul li.user {
  color: #2D67F6;
  font-size: 1.3333333333333333rem;
  font-weight: bold;
  margin: 0.5rem 1rem;
}

.c3 > div ul > b {
  margin: 0.5rem 1rem;
  position: relative;
  width: -webkit-min-content;
  width: -moz-min-content;
  width: min-content;
}

.c3 > div ul > b::after {
  content: '';
  position: absolute;
  opacity: calc(0.5);
  top: 50%;
  -webkit-transform: translate(100%,-50%);
  -ms-transform: translate(100%,-50%);
  transform: translate(100%,-50%);
  right: -0.5rem;
  width: 4.583333333333333rem;
  height: 1px;
  background-color: #4D4D4D;
}

.c3 > div ul > li {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.c3 > div ul > li a,
.c3 > div ul > li button {
  width: 100%;
  padding: 0.6666666666666666rem 1.25rem 0.6666666666666666rem 0.8333333333333334rem;
  font-size: 1.5rem;
  text-align: left;
}

.c3 > div ul > li a:hover,
.c3 > div ul > li button:hover {
  background-color: #ed225d;
  color: #FFF;
}

<header>
    <nav
      class="c0"
    >
      <div
        class="c1"
      >
        <test-file-stub />
      </div>
      <div
        class="c2"
      >
        <h1>
          <span
            class="editable-input editable-input--is-not-editing editable-input--has-value "
          >
            <button
              aria-hidden="false"
              aria-label="Edit sketch name"
              class="editable-input__label"
            >
              <span>
                Test project name
              </span>
              <test-file-stub
                aria-hidden="true"
                classname="editable-input__icon"
                focusable="false"
              />
            </button>
            <input
              aria-hidden="true"
              aria-label="New sketch name"
              class="editable-input__input"
              disabled=""
              maxlength="128"
              type="text"
              value="Test project name"
            />
          </span>
        </h1>
      </div>
      <div
        class="c3"
      >
        <div>
          <a
            href="/login"
          >
            <button
              class="c4 c5"
              display="inline"
              focusable="false"
              kind="primary"
              type="button"
            >
              <test-file-stub
                aria-hidden="true"
                classname="sc-fujyAs cSTVlM"
                focusable="false"
              />
            </button>
          </a>
        </div>
        <div>
          <button
            class="c4 c5"
            display="inline"
            focusable="false"
            kind="primary"
            type="button"
          >
            <test-file-stub
              aria-hidden="true"
              classname="sc-iqAclL iOZiVo"
              focusable="false"
            />
          </button>
          <ul
            class=""
          >
            <b>
              File
            </b>
            <li
              class="nav__dropdown-item"
            >
              <button>
                New
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Save
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <a
                href="/p5/sketches"
              >
                Examples
              </a>
            </li>
            <b>
              Edit
            </b>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Tidy Code
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Find
              </button>
            </li>
            <b>
              Sketch
            </b>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Add File
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Add Folder
              </button>
            </li>
            <b>
              Settings
            </b>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Preferences
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Language
              </button>
            </li>
            <b>
              Help
            </b>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Keyboard Shortcuts
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <a
                href="https://p5js.org/reference/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Reference
              </a>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <a
                href="/about"
              >
                About
              </a>
            </li>
          </ul>
        </div>
      </div>
    </nav>
  </header>
</DocumentFragment>
`;

exports[`Nav renders editor version for desktop 1`] = `
<DocumentFragment>
  <header>
    <nav
      class="nav"
    >
      <ul
        class="nav__items-left"
      >
        <li
          class="nav__item-logo"
        >
          <test-file-stub
            aria-label="p5.js Logo"
            classname="svg__logo"
            focusable="false"
            role="img"
          />
        </li>
        <li
          class="nav__item"
        >
          <button>
            <span
              class="nav__item-header"
            >
              File
            </span>
            <test-file-stub
              aria-hidden="true"
              classname="nav__item-header-triangle"
              focusable="false"
            />
          </button>
          <ul
            class="nav__dropdown"
          >
            <li
              class="nav__dropdown-item"
            >
              <button>
                New
              </button>
            </li>
          </ul>
        </li>
        <li
          class="nav__item"
        >
          <button>
            <span
              class="nav__item-header"
            >
              Edit
            </span>
            <test-file-stub
              aria-hidden="true"
              classname="nav__item-header-triangle"
              focusable="false"
            />
          </button>
          <ul
            class="nav__dropdown"
          >
            <li
              class="nav__dropdown-item"
            >
              <button>
                Tidy Code
                <span
                  class="nav__keyboard-shortcut"
                >
                  ⌃+⇧+F
                </span>
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Find
                <span
                  class="nav__keyboard-shortcut"
                >
                  ⌃+F
                </span>
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Replace
                <span
                  class="nav__keyboard-shortcut"
                >
                  ⌃+H
                </span>
              </button>
            </li>
          </ul>
        </li>
        <li
          class="nav__item"
        >
          <button>
            <span
              class="nav__item-header"
            >
              Sketch
            </span>
            <test-file-stub
              aria-hidden="true"
              classname="nav__item-header-triangle"
              focusable="false"
            />
          </button>
          <ul
            class="nav__dropdown"
          >
            <li
              class="nav__dropdown-item"
            >
              <button>
                Add File
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Add Folder
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Run
                <span
                  class="nav__keyboard-shortcut"
                >
                  ⌃+Enter
                </span>
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Stop
                <span
                  class="nav__keyboard-shortcut"
                >
                  ⇧+⌃+Enter
                </span>
              </button>
            </li>
          </ul>
        </li>
        <li
          class="nav__item"
        >
          <button>
            <span
              class="nav__item-header"
            >
              Help
            </span>
            <test-file-stub
              aria-hidden="true"
              classname="nav__item-header-triangle"
              focusable="false"
            />
          </button>
          <ul
            class="nav__dropdown"
          >
            <li
              class="nav__dropdown-item"
            >
              <button>
                Keyboard Shortcuts
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <a
                href="https://p5js.org/reference/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Reference
              </a>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <a
                href="/about"
              >
                About
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </nav>
  </header>
</DocumentFragment>
`;

exports[`Nav renders editor version for mobile 1`] = `
<DocumentFragment>
  .c4.c4.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #333;
  cursor: pointer;
  border: none;
  line-height: 1;
}

.c4.c4.c4 svg * {
  fill: #333;
}

.c4.c4.c4:disabled {
  cursor: not-allowed;
}

.c4.c4.c4 > * + * {
  margin-left: 0.6666666666666666rem;
}

.c5 {
  width: 4rem;
}

.c5 > svg {
  width: 100%;
  height: 100%;
}

.c0 {
  background: #F0F0F0;
  color: #333;
  padding: 0.6666666666666666rem 0;
  gap: 1.1666666666666667rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.8333333333333334rem;
  box-shadow: #00000030 0px 2px 8px 0px;
  z-index: 10;
}

.c1 {
  width: 2.3333333333333335rem;
  aspect-ratio: 1;
  margin-left: 1.1666666666666667rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 > svg {
  width: 100%;
  height: 100%;
}

.c1 > svg > path {
  fill: #ed225d;
  stroke: #ed225d;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.16666666666666666rem;
}

.c2 * {
  padding: 0;
  margin: 0;
}

.c2 > h5 {
  font-size: 1.0833333333333333rem;
  font-weight: normal;
}

.c3 {
  margin-left: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 ul.opened {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  opacity: 1;
  max-width: 90vw;
  max-height: 90vh;
  overflow-y: auto;
}

.c3 > div {
  position: relative;
}

.c3 > div ul {
  foreground: #000;
  background: #F0F0F0;
  border: #A6A6A6;
  position: absolute;
  overflow: hidden;
  z-index: 10;
  right: 10px;
  -webkit-transform: translateX(0.5rem);
  -ms-transform: translateX(0.5rem);
  transform: translateX(0.5rem);
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  opacity: 0;
  -webkit-transform-origin: top;
  -ms-transform-origin: top;
  transform-origin: top;
  -webkit-transition: opacity 100ms,-webkit-transform 100ms 200ms;
  -webkit-transition: opacity 100ms,transform 100ms 200ms;
  transition: opacity 100ms,transform 100ms 200ms;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.16666666666666666rem;
  box-shadow: rgba(60,64,67,0.3) 0px 1px 2px 0px, rgba(60,64,67,0.15) 0px 1px 3px 1px;
  min-width: 10rem;
  border-radius: 0.16666666666666666rem;
}

.c3 > div ul li.user {
  color: #2D67F6;
  font-size: 1.3333333333333333rem;
  font-weight: bold;
  margin: 0.5rem 1rem;
}

.c3 > div ul > b {
  margin: 0.5rem 1rem;
  position: relative;
  width: -webkit-min-content;
  width: -moz-min-content;
  width: min-content;
}

.c3 > div ul > b::after {
  content: '';
  position: absolute;
  opacity: calc(0.5);
  top: 50%;
  -webkit-transform: translate(100%,-50%);
  -ms-transform: translate(100%,-50%);
  transform: translate(100%,-50%);
  right: -0.5rem;
  width: 4.583333333333333rem;
  height: 1px;
  background-color: #4D4D4D;
}

.c3 > div ul > li {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.c3 > div ul > li a,
.c3 > div ul > li button {
  width: 100%;
  padding: 0.6666666666666666rem 1.25rem 0.6666666666666666rem 0.8333333333333334rem;
  font-size: 1.5rem;
  text-align: left;
}

.c3 > div ul > li a:hover,
.c3 > div ul > li button:hover {
  background-color: #ed225d;
  color: #FFF;
}

<header>
    <nav
      class="c0"
    >
      <div
        class="c1"
      >
        <test-file-stub />
      </div>
      <div
        class="c2"
      >
        <h1>
          <span
            class="editable-input editable-input--is-not-editing editable-input--has-value "
          >
            <button
              aria-hidden="false"
              aria-label="Edit sketch name"
              class="editable-input__label"
            >
              <span>
                Test project name
              </span>
              <test-file-stub
                aria-hidden="true"
                classname="editable-input__icon"
                focusable="false"
              />
            </button>
            <input
              aria-hidden="true"
              aria-label="New sketch name"
              class="editable-input__input"
              disabled=""
              maxlength="128"
              type="text"
              value="Test project name"
            />
          </span>
        </h1>
      </div>
      <div
        class="c3"
      >
        <div>
          <a
            href="/login"
          >
            <button
              class="c4 c5"
              display="inline"
              focusable="false"
              kind="primary"
              type="button"
            >
              <test-file-stub
                aria-hidden="true"
                classname="sc-fujyAs cSTVlM"
                focusable="false"
              />
            </button>
          </a>
        </div>
        <div>
          <button
            class="c4 c5"
            display="inline"
            focusable="false"
            kind="primary"
            type="button"
          >
            <test-file-stub
              aria-hidden="true"
              classname="sc-iqAclL iOZiVo"
              focusable="false"
            />
          </button>
          <ul
            class=""
          >
            <b>
              File
            </b>
            <li
              class="nav__dropdown-item"
            >
              <button>
                New
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Save
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <a
                href="/p5/sketches"
              >
                Examples
              </a>
            </li>
            <b>
              Edit
            </b>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Tidy Code
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Find
              </button>
            </li>
            <b>
              Sketch
            </b>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Add File
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Add Folder
              </button>
            </li>
            <b>
              Settings
            </b>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Preferences
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Language
              </button>
            </li>
            <b>
              Help
            </b>
            <li
              class="nav__dropdown-item"
            >
              <button>
                Keyboard Shortcuts
              </button>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <a
                href="https://p5js.org/reference/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Reference
              </a>
            </li>
            <li
              class="nav__dropdown-item"
            >
              <a
                href="/about"
              >
                About
              </a>
            </li>
          </ul>
        </div>
      </div>
    </nav>
  </header>
</DocumentFragment>
`;
